src/Microsoft.ML.Recommend/MatrixFactorizationPredictor.cs

```diff
@@ -0,0 +1,403 @@
+            private readonly string _xColName;
+            private readonly string _yColName;
+
+            public ISchema InputSchema => _inputSchema.Schema;
+
+            public RoleMappedSchema InputRoleMappedSchema => _inputSchema;
+
+            public RowMapper(MatrixFactorizationPredictor parent, RoleMappedSchema schema, ScoreMapperSchema outputSchema)
+            {
+                Contracts.AssertValue(parent);
+                _parent = parent;
+
+                var list = schema.GetColumns(RecommendUtils.XKind);
+                if (Utils.Size(list) != 1)
+                    throw Contracts.Except($"'{RecommendUtils.XKind}' column doesn't exist");
+                _xColName = list[0].Name;
+                _xColIndex = list[0].Index;
+
+                list = schema.GetColumns(RecommendUtils.YKind);
+                if (Utils.Size(list) != 1)
+                    throw Contracts.Except($"'{RecommendUtils.YKind}' column doesn't exist");
+                _yColName = list[0].Name;
+                _yColIndex = list[0].Index;
+
+                CheckInputSchema(schema.Schema, _xColIndex, _yColIndex);
+                _inputSchema = schema;
+                _outputSchema = outputSchema;
+            }
+
+            public ISchema Schema => OutputSchema;
+
+            public ISchema OutputSchema
+            {
+                get { return _outputSchema; }
+            }
+
+            public Func<int, bool> GetDependencies(Func<int, bool> predicate)
+            {
+                for (int i = 0; i < OutputSchema.ColumnCount; i++)
+                {
+                    if (predicate(i))
+                        return col => (col == _xColIndex || col == _yColIndex);
+                }
+                return col => false;
+            }
+
+            public IEnumerable<KeyValuePair<RoleMappedSchema.ColumnRole, string>> GetInputColumnRoles()
+            {
+                yield return RecommendUtils.XKind.Bind(_xColName);
+                yield return RecommendUtils.YKind.Bind(_yColName);
+            }
+
+            private void CheckInputSchema(ISchema schema, int xCol, int yCol)
+            {
+                var type = schema.GetColumnType(xCol);
+                if (!type.Equals(_parent.InputXType))
+                {
+                    throw Contracts.ExceptParam(nameof(schema), "Input X type '{0}' incompatible with predictor X type '{1}'",
+                        type, _parent.InputXType);
+                }
+                type = schema.GetColumnType(yCol);
+                if (!type.Equals(_parent.InputYType))
+                {
+                    throw Contracts.ExceptParam(nameof(schema), "Input Y type '{0}' incompatible with predictor Y type '{1}'",
+                        type, _parent.InputYType);
+                }
+            }
+
+            private Delegate[] CreateGetter(IRow input, bool[] active)
+            {
+                Contracts.CheckValue(input, nameof(input));
```

Comment:
>Contracts [](start = 16, length = 9)

Should these also use the environment from the parent?

Reply:
Thanks. I just add a private property to this RowMapper:
            private IHost Host => _parent._host;
Then, all checks will be invoked via something like Host.CheckValue(...).

---
In reply to: [225356214](https://github.com/dotnet/machinelearning/pull/1263#discussion_r225356214) [](ancestors = 225356214)